---
title: "WT SDA"
output: html_notebook
---

This is running SDA only on WT cell (no Knock Outs). In general, we find many component are very similar across the different SDA decompositions.

```{r}
library(Matrix)
library(data.table)
library(dropsim)

raw_data <- readRDS("../data/merged-mouse_counts_tsne-QC_low-mt.rds")
raw_data <- t(raw_data)

datat <- readRDS("../data/tmp_cache/datat.rds")

wt_groups <- c("WT","mj","SPG","SPD","SPCII","SPCI")

str(wt_groups)

wt_cells <- datat[group %in% wt_groups]$cell



raw_data <- raw_data[,wt_cells]

data <- normaliseDGE(raw_data,
                     center = FALSE,
                     scale = TRUE,
                     threshold = 10,
                     min_library_size = 200,
                     gene_subset = (2/3))

library_size <- Matrix::colSums(raw_data[,rownames(data)])

# Issue warning if data changes
stopifnot(digest::digest(data) == "9996e625cd14c3387d0840442de901ba")
stopifnot(digest::digest(library_size) == "21c2f00d2a33fff3dab0082b78d179b2")

saveRDS(list(dge=data, library_size=library_size), "../data/merged_mouse_normalised_tsne-QC_V3_WT.rds") # save a cache for easy reopening
```

```{r}
data <- readRDS("../data/conradV3/merged_mouse_normalised_tsne-QC_V3_WT.rds")$dge
export_data(as.matrix(data), name = "merged_mouse_V3_WT_SDA", path="../data/conradV3/")

run_SDA(sda_location = "../../SDA/build/sda",
        out = "../results/conradV3_sda_9",
        data = "../data/conradV3/merged_mouse_V3_WT_SDA.data",
        num_comps = 50,
        max_iter = 10000,
        save_freq = 1000,
        set_seed = "79151 17351",
        N = 9936,
        eigen_parallel = TRUE,
        num_blocks = 5,
        num_openmp_threads = 8)

```




        
```{r}
library(SDAtools)
library(ggplot2)
library(ggrepel)
library(data.table)

results <- load_results(results_folder = "../data/SDA/conradV3_sda_9/", data_path = "../data/count_matrices/")
str(results)

rownames(results$loadings[[1]]) <- paste0("WT_",1:50)
colnames(results$scores) <- paste0("WT_",1:50)

check_convergence(results)
loading_distribution(results)
scores_distribution(results)
plot_maximums(results)
#plot_scree(results)
PIP_distribution(results)
PIP_component_distribution(results, 2)
PIP_threshold_distribution(results)
```

```{r}

rna_locations <- load_gene_locations(colnames(results$loadings[[1]]), name="merge_mouse_V3_WT", path = "../data/")
rna_locations[,.N,by=chromosome_name][order(-N)]

```

# Overall Comparison

Generally correlations between any random compoents are low (good), with some pairs have high correlation (~0.8, good). Permuting the gene loadings gives a kind of null distribution of correlations which are all below 0.03. Most (11) WT components have one good match (>0.5) to the mix compoents, with 1 having two matches, and 38 no match.

```{r, fig.width=6}
results1 <- load_results(results_folder = "../data/SDA/conradV3_sda_1/", data_path = "../data/count_matrices/")
str(results1)

rownames(results1$loadings[[1]]) <- paste0("Mix_",1:50)

library(ComplexHeatmap)

common_genes <- colnames(results1$loadings[[1]])[colnames(results1$loadings[[1]]) %in% colnames(results$loadings[[1]])]

cross_cor <- abs(cor(t(results1$loadings[[1]][,common_genes]),
                 t(results$loadings[[1]][,common_genes]),
                 method = "spearman"))

cross_cor_rand <- abs(cor(t(results1$loadings[[1]][sample(1:50),sample(common_genes)]),
                 t(results$loadings[[1]][,common_genes]),
                 method = "spearman"))

dt_long <- data.table(id=1:50^2, correlation=c(cross_cor), permuted_null_correlation=c(cross_cor_rand))

ggplot(melt(dt_long, id.vars = "id"), aes(value)) +
        geom_histogram(bins=1000) +
        facet_wrap(~variable, scales = "free_y", nrow = 2) 


max_cor_per_compoent <- sort(apply(cross_cor, 2, max), decreasing = T)
plot(max_cor_per_compoent)

table(apply(cross_cor, 2, function(x) sum(x>0.5)))
```

Using heatmap of the correlation between components from WT only and mixed KO & WT we can see many pairs with high correlations with 1 or sometimes 2 components. There's quite a few with low correlations because I ran SDA with 50 components even with WT only to be fair, but there's less complexity with just WT cells so we're getting more uninteresting components / single cell (38 components have a max score > 30 vs only 11 in full (mixed WT&KO) analysis).

```{r, fig.width=8}
mix_comps <- rownames(cross_cor)

mix_order <- character()

for (i in seq_along(max_cor_per_compoent)){
        tmp2 <- names(which.max(cross_cor[mix_comps, names(max_cor_per_compoent)[i] ]))
        if(is.null(tmp2)){
                tmp2 <- mix_comps
        }
        mix_order <- c(mix_order, tmp2)
        mix_comps <- mix_comps[!mix_comps %in% tmp2]
}

tmp2 <- data.table(score_sum_abs = colMeans(abs(results$scores))[names(max_cor_per_compoent)],
                   cells_in_component = colSums(abs(results$scores)>1)[names(max_cor_per_compoent)],
                   score_sum_2 = colMeans(results$scores^2)[names(max_cor_per_compoent)],
                   score_sum_med_2 = apply(results$scores^2, 2, function(x) quantile(x, probs=0.98))[names(max_cor_per_compoent)],
                   max_cor_per_compoent,
                   max_cell_score = apply(abs(results$scores), 2, max)[names(max_cor_per_compoent)],
                   component=names(max_cor_per_compoent))

library(circlize)

ha = rowAnnotation(df = tmp2[,.(cells_in_component, max_cell_score)],
                   col = list(cells_in_component=colorRamp2(c(0, max(tmp2$cells_in_component)), c("white", "red")),
                              max_cell_score=colorRamp2(c(0, 95), c("white", "blue"))),
                   annotation_legend_param=list(legend_direction = "horizontal",
                                                legend_width = unit(5, "cm"), title_position = "lefttop")
                   )

#, col = list(score_sum_count = colorRamp2(quantile(tmp2$score_sum_count, seq(0, 1, by = 0.25)), viridis(5)),
                                                                     #max_score = colorRamp2(quantile(tmp2$max_score, seq(0, 1, by = 0.25)), viridis(5,direction = -1)))

#max_score = viridisLite::viridis(100), 
#colorRamp2(c(0, 20), c("white", "red"))

gene_loading_correlation = t(cross_cor[mix_order, names(max_cor_per_compoent)][50:1,])

hm = Heatmap(gene_loading_correlation,
        col = viridisLite::viridis(100),
        cluster_rows = F,
        cluster_columns = F,
        row_title = "Components from WT only",
        column_title = "Components from WT & KO Combined",
        name = "Gene loading correlation",
        heatmap_legend_param = list(legend_direction = "horizontal",
                                    legend_width = unit(5, "cm"), title_position = "lefttop"))

pdf(width = 12, height = 10, file = "../results/WT_vs_Mixed.pdf")
draw(hm + ha, heatmap_legend_side = "bottom")
dev.off()

draw(hm + ha, heatmap_legend_side = "bottom")
```

The component with high sum of squared scores (high contribution) have high correlations. Also those with high max score (indicating ovefitting with a single high cell in one component) have low correlations.

```{r, fig.width=6}



ggplot(tmp2, aes(score_sum_2, max_cor_per_compoent, label=component)) +
        geom_point() + 
        geom_label_repel(data=tmp2[max_cor_per_compoent>0.4])

ggplot(tmp2, aes(max_score, max_cor, label=component)) +
        geom_point() + 
        geom_label_repel(data=tmp2[max_cor>0.4])

# ggplot(tmp2, aes(score_sum_abs, max_cor, label=component)) +
#         geom_point() + 
#         geom_label_repel(data=tmp2[max_cor>0.4])
# 
ggplot(tmp2, aes(score_sum_count, max_cor_per_compoent, label=component)) +
        geom_point() +
        geom_label_repel(data=tmp2[max_cor_per_compoent>0.4])

# Heatmap(abs(cor(cbind(t(results1$loadings[[1]][,common_genes]),
#                       t(results$loadings[[1]][,common_genes])))),
#         col = viridisLite::viridis(100))

# Heatmap(cross_cor,
#         col = viridisLite::viridis(100),
#         clustering_distance_rows = "maximum",
#         clustering_distance_columns = "maximum")
```

# tSNE

This is tSNE on *all* the components (no batch effects removed).

The marker genes pachytene-onwards still look good. Prior to that we don't have many cells to look at.

```{r, fig.width=10}
set.seed(42)
tsne_data_wt <- Rtsne::Rtsne(results$scores, verbose=TRUE, pca=FALSE, perplexity = 30, max_iter=1000)

marker_genes <- c("Csf1r", "Dcn","Cyp11a1","Aard","Trim7","Gfra1","Nanos1","Esx1","Ctcfl","Prdm9","Ccnb3","Piwil1","Ccna1","Ssxb1","Acrv1","Lrrd1","Fam71b","Tnp2","Prm2","mt-Rnr2")
imputed_expression <- results$scores %*% results$loadings[[1]][,marker_genes]

datat <- data.table(cell_id=1:9936, tsne_data_wt$Y, imputed_expression)
setnames(datat, c("V1","V2"),c("tSNE1","tSNE2"))
marker_plot_list <- list()

print_wt_marker <- function(i){
p <- ggplot(datat, aes(tSNE1, tSNE2, colour=get(i))) +
        geom_point(size=0.5) +
        scale_color_viridis(direction = -1) + 
        ggtitle(i)
return(p)
}

library(testisAtlas)
grid.arrange(grobs=create_grob_list(fn = print_wt_marker, input = marker_genes), nrow=4, heights = c(1,1,1,1))

save(results, datat, cross_cor, file = "../data/tmp_cache/SDA_WT.rds")

```


## Mj - 15

Component 15 has an excellent match with mix component 6 (Mj) 

```{r}

plot(results1$loadings[[1]][6,common_genes],results$loadings[[1]][15,common_genes])
cor(results1$loadings[[1]][6,common_genes],results$loadings[[1]][15,common_genes])

genome_loadings(results$loadings[[1]]["WT_15",])
```

## Round Spermatid - 35

Component 35 has an excellent match with mix component 15 (round Spermatid)

```{r}

plot(results1$loadings[[1]][15,common_genes],results$loadings[[1]][35,common_genes])
cor(results1$loadings[[1]][15,common_genes],results$loadings[[1]][35,common_genes])

genome_loadings(results$loadings[[1]]["WT_35",])
```

## Acrosomal - 40

Component 40 has high match with mix component 30 (Acrosomal) Spaca1, Sox5, Lyzl6, Lyzl1, Spata31, Acrv1, Tex36 etc.

```{r}

plot(results1$loadings[[1]][30,common_genes],results$loadings[[1]][40,common_genes])
cor(results1$loadings[[1]][30,common_genes],results$loadings[[1]][40,common_genes])

genome_loadings(results$loadings[[1]]["WT_40",])
```

## Pachytene - 42

Lot's of high genes in common with the pachytene components 42 and 39 from KO&WT mixed SDA - Ldhc, Pabpc1, Lyar, Ybx3, Nasp, Dkk1, Ropn1l, Phf7, Fabp9 etc.

```{r}
plot(results1$loadings[[1]][42,common_genes],results$loadings[[1]][42,common_genes])
plot(results1$loadings[[1]][39,common_genes],results$loadings[[1]][42,common_genes])
cor(results1$loadings[[1]][42,common_genes],results$loadings[[1]][42,common_genes])
cor(results1$loadings[[1]][39,common_genes],results$loadings[[1]][42,common_genes])
genome_loadings(results$loadings[[1]]["WT_42",])
```



##Differentiated Spermatogonia - 28

High correlation with component 33 (diff spermatogonia) Dmrt1, Uchl1, Dazl, Mapd72

```{r}
plot(results1$loadings[[1]][33,common_genes],results$loadings[[1]][28,common_genes])
cor(results1$loadings[[1]][33,common_genes],results$loadings[[1]][28,common_genes])
genome_loadings(results$loadings[[1]]["WT_28",])
```

## Spermiogenesis - 7

High correlation with component 17 from KO&WT mix (Spermiogenesis) - Prm1, Tnp2, Tnp1

```{r}
plot(results1$loadings[[1]][17,common_genes],results$loadings[[1]][7,common_genes])
cor(results1$loadings[[1]][17,common_genes],results$loadings[[1]][7,common_genes])
genome_loadings(results$loadings[[1]]["WT_7",])
```

## Spermiogenesis - 46

Component 46 has high concordance with mixed component 18 & 17 - spermiogenesis (Prm2, Oaz3, Smcp, Gstm5, Dbil5, Tppp2, Gapdhs, Gsg1, Tssk6 etc.)

```{r}
plot(results1$loadings[[1]][18,common_genes],results$loadings[[1]][46,common_genes])
cor(results1$loadings[[1]][18,common_genes],results$loadings[[1]][46,common_genes])
plot(results1$loadings[[1]][17,common_genes],results$loadings[[1]][46,common_genes])
cor(results1$loadings[[1]][17,common_genes],results$loadings[[1]][46,common_genes])
genome_loadings(results$loadings[[1]]["WT_46",])
```

## Leydig - 17

Component 17 has high match with 40 & 24 (Leydig) Insl3, Cyp17a1, Aldh1a1, Agt, Mgst1, Ptgds, Lcn2, Akr1cl, Cst3, Gstm1, Fabp3 etc.

```{r}
plot(results1$loadings[[1]][24,common_genes],results$loadings[[1]][17,common_genes])
plot(results1$loadings[[1]][40,common_genes],results$loadings[[1]][17,common_genes])
cor(results1$loadings[[1]][24,common_genes],results$loadings[[1]][17,common_genes])
cor(results1$loadings[[1]][40,common_genes],results$loadings[[1]][17,common_genes])
genome_loadings(results$loadings[[1]]["WT_17",])
```

## Respiration - 23

Component 23 has good match with mix component 9 respiration

```{r}

plot(results1$loadings[[1]][9,common_genes],results$loadings[[1]][23,common_genes])
cor(results1$loadings[[1]][9,common_genes],results$loadings[[1]][23,common_genes])

genome_loadings(results$loadings[[1]]["WT_23",])
```

## Fibroblast - 39

Component 39 has high correlaiton with Mix component 32 (peritubular myoid / fibroblast)

```{r}
plot(results1$loadings[[1]][32,common_genes],results$loadings[[1]][39,common_genes])
cor(results1$loadings[[1]][32,common_genes],results$loadings[[1]][39,common_genes])
genome_loadings(results$loadings[[1]]["WT_39",])
```

## Post division/acrosomal - 16

Component 16 is a good match to the positive side of mixed component 35

```{r}
plot(results1$loadings[[1]][35,common_genes],results$loadings[[1]][16,common_genes])
cor(results1$loadings[[1]][35,common_genes],results$loadings[[1]][16,common_genes])
genome_loadings(results$loadings[[1]]["WT_16",])
```



## Undiff Spermatogonia - 18

Component 18 has high match with 31 (undiff spermatogonia) Egr4, Zbtb16, Sdc4, Afp, Foxo1, Erap1, Tubb5, Fbxo32 etc.

```{r}
plot(results1$loadings[[1]][31,common_genes],results$loadings[[1]][18,common_genes])

mix_wt <- data.table(Mix_C31_loading = results1$loadings[[1]][31,common_genes], WT_only_C18_loading = results$loadings[[1]][18,common_genes], gene=common_genes)

pdf(width = 20, height = 10, file = "../results/mix_wt_example.pdf")
plot_grid(NULL,
          ggplot(mix_wt, aes(Mix_C31_loading, WT_only_C18_loading, label=gene)) +
    geom_point(alpha=0.3) +
    theme_minimal() +
    geom_text_repel(data=mix_wt[abs(Mix_C31_loading)>0.19 | abs(WT_only_C18_loading)>0.2], force = 2),
    labels = "AUTO")
dev.off()

cor(results1$loadings[[1]][31,common_genes],results$loadings[[1]][18,common_genes])
genome_loadings(results$loadings[[1]]["WT_18",])
```





## Stem Cells - 32

Component 32 matches ok with mix componet 50 stem cell - Glis3, Sox11, Dusp6, Mcam, H19, Ccnd2, 

```{r}

plot(results1$loadings[[1]][50,common_genes],results$loadings[[1]][32,common_genes])
cor(results1$loadings[[1]][50,common_genes],results$loadings[[1]][32,common_genes])

genome_loadings(results$loadings[[1]]["WT_32",])
```



## Sertoli - 30

Component 30 has ok match to mixed sertoli 45 (Aard, Ldhb, Cst12) & sertoli/muscle 10 (Kazald, Jun, Cyr61)

```{r}

plot(results1$loadings[[1]][45,common_genes],results$loadings[[1]][30,common_genes])
plot(results1$loadings[[1]][10,common_genes],results$loadings[[1]][30,common_genes])
cor(results1$loadings[[1]][45,common_genes],results$loadings[[1]][30,common_genes])
cor(results1$loadings[[1]][10,common_genes],results$loadings[[1]][30,common_genes])

genome_loadings(results$loadings[[1]]["WT_30",])
```





## Early Pachytene - 49

Component 49 has high match with 13 (Early Pachy) Piwil1, 

```{r}

plot(results1$loadings[[1]][13,common_genes],results$loadings[[1]][49,common_genes])
cor(results1$loadings[[1]][13,common_genes],results$loadings[[1]][49,common_genes])

genome_loadings(results$loadings[[1]]["WT_49",])
```

